manpages/makerank.1
.\" -*- mode: nroff; -*-
.Dd March 15, 2024
.Dt MAKERANK 1
.Os
.Sh NAME
.Nm makerank
.Nd creates rankings for FICS games
.Sh SYNOPSIS
Takes no arguments.
.Sh DESCRIPTION
.Nm
creates sorted rankings for FICS (Free Internet Chess Server) games
and writes them to the hard disk.
Currently for the following game types:
.Bl -bullet -compact
.It
std
.It
blitz
.It
wild
.It
lightning
.El
.Sh SEE ALSO
.Xr fics 1 , Xr fics_addplayer 1


CHANGELOG.md
# Change Log #

All notable changes to this fork of FICS version 1.6.2 will be
documented in this file.

## [Unreleased] ##
- Added `PRINTFLIKE()` and fixed many format errors.
- Added argument lists to many function declarations.
- Added initialization of many variables.
- Added newly written manual pages.
- Added usage of `ARRAY_SIZE()`
- Changed the addplayer program to generate passwords that are 8
  characters long (was 4).
- Deleted non-existent functions from the header files.
- Deleted obsolete and unused code
- Fixed format strings that weren't string literals. Potentially
  insecure.
- Fixed incorrect buffer sizes
- Fixed unchecked return values
- Made functions and variables private (aka static) where possible.
- Marked functions that doesn't return `__dead`.
- Redefined `ASSERT()`
- Reformatted code according to OpenBSD's KNF.
- Replaced `rand()` calls with arc4random.
- Replaced _multiple_ `sprintf()` calls with `snprintf()` + truncation
  checks.
- Replaced _multiple_ occurrences of `strcpy()` and `strcat()` with
  size-bounded versions.

## [1.0] - 2023-12-28 ##
- Added a new build system (and deleted the old).
- Added argument lists to many function declarations.
- Added auto-generation of the header file `ficspaths.h` and included
  it in `config.h`.
- Added better handling of memory allocation errors.
- Added usage of the `time_t` typedef in multiple places. This instead
  of `int`.
- Declared file-local functions and variables as `PRIVATE`.
- Deleted disabled code
- Did new revisions of the following files:
  - `board.c`
  - `eco.c`
  - `fics_addplayer.c`
  - `formula.c`
  - `makerank.c`
  - `network.c`
  - `shutdown.c`
  - `stdinclude.h`
  - `utils.c`
  - ...
- Fixed a `sscanf()` bug in `com_anews()`.
- Fixed a bug in `fix_time()` (did return a local address)
- Fixed bogus `crypt()` calls. (The second arg was wrong.)
- Fixed bugs in `process_move()`
- Fixed bugs in `stored_mail_moves()`
- Fixed comparison between pointer and integer in `com_inchannel()`.
- Fixed dead assignments
- Fixed implicit integers
- Fixed _multiple_ cases of use of possibly uninitialized variables.
- Fixed _multiple_ possible buffer overflows
- Fixed _multiple_ `sprintf()` format overflows
- Fixed the type of the variable passed to `strgtime()` and
  `strltime()`. Multiple occurrences.
- Fixed unused variables
- Reformatted code according to OpenBSD's KNF.
- Renamed functions in order to avoid conflicts with system
  declarations.

## 2023-12-07 ##
- Forked FICS version 1.6.2 made by Richard Nash.


FICS/makerank.c
/*
 * Revised by maxxe 23/12/14
 */

#include <ctype.h>
#include <err.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#if __linux__
#include <bsd/string.h>
#endif

#include "common.h"
#include "makerank.h"

static ENTRY	**list;
static ENTRY	**sortme;

static char	*rnames[] = { "std", "blitz", "wild", "lightning" };
static int	 rtype;

static int
GetPlayerInfo(char *fileName, ENTRY *e)
{
	FILE	*fp;
	char	 NameWithCase[30];
	char	 field[20];
	char	 line[100];
	int	 i, done = 0;

	e->computer = 0;

	for (i = 0; i < ARRAY_SIZE(e->r); i++) {
		e->r[i].num	= 0;
		e->r[i].rating	= 0;
	}

	if ((fp = fopen(fileName, "r")) == NULL ||
	    fgets(line, sizeof line - 1, fp) == NULL ||
	    feof(fp))
		return 0;

	if (!strcmp(line, "v 1\n")) {
		fgets(line, 99, fp);

		sscanf(line, "%s", e->name);

		fgets(line, 99, fp);
		fgets(line, 99, fp);
		fgets(line, 99, fp);

		if (fscanf(fp, "%u %*u %*u %*u %u %*u %*u %*u %*u %u %*u %*u "
		    "%*u %u %*u %*u %*u %*u %u %*u %*u %*u %u %*u %*u %*u %*u "
		    "%u %*u %*u %*u %u %*u %*u %*u %*u",
		    &(e->r[0].num),
		    &(e->r[0].rating),
		    &(e->r[1].num),
		    &(e->r[1].rating),
		    &(e->r[2].num),
		    &(e->r[2].rating),
		    &(e->r[3].num),
		    &(e->r[3].rating)) != 8) {
			fprintf(stderr, "OOPS: couldn't parse player file %s."
			    "\n", fileName);
		}
	} else {
		do {
			sscanf(line, "%s", field);

			if (!strcmp(field, "Name:")) {
				sscanf(line, "%*s %s", NameWithCase);

				if (strcasecmp(e->name, NameWithCase)) {
					printf("TROUBLE: %s's handle is "
					    "listed as %s.\n", e->name,
					    NameWithCase);
				} else if (strlcpy(e->name, NameWithCase,
				    sizeof e->name) >= sizeof e->name) {
					fprintf(stderr, "%s: warning: "
					    "strlcpy() truncated\n", __func__);
				}
			} else if (!strcmp(field, "S_NUM:")) {
				sscanf(line, "%*s %d", &(e->r[0].num));
			} else if (!strcmp(field, "B_NUM:")) {
				sscanf(line, "%*s %d", &(e->r[1].num));
			} else if (!strcmp(field, "W_NUM:")) {
				sscanf(line, "%*s %d", &(e->r[2].num));
			} else if (!strcmp(field, "L_NUM:")) {
				sscanf(line, "%*s %d", &(e->r[3].num));
			} else if (!strcmp(field, "S_RATING:")) {
				sscanf(line, "%*s %d", &(e->r[0].rating));
			} else if (!strcmp(field, "B_RATING:")) {
				sscanf(line, "%*s %d", &(e->r[1].rating));
			} else if (!strcmp(field, "W_RATING:")) {
				sscanf(line, "%*s %d", &(e->r[2].rating));
			} else if (!strcmp(field, "L_RATING:")) {
				sscanf(line, "%*s %d", &(e->r[3].rating));
			} else if (!strcmp(field, "Network:")) {
				done = 1;
			}

			fgets(line, 99, fp);
		} while (!done && !feof(fp));
	}

	fclose(fp);
	return 1;
}

static int
LoadEntries(void)
{
	ENTRY	 e;
	FILE	*fpPlayerList;
	char	 command[90];
	char	 letter1;
	char	 pathInput[80];
	int	 len, n = 0;
	int	 listsize;

	listsize	= 100;
	list		= malloc(sizeof(ENTRY *) * listsize);

	if (list == NULL)
		err(1, "%s: malloc", __func__);

	for (letter1 = 'a'; letter1 <= 'z'; letter1++) {
		printf("Loading %c's.\n", letter1);

		snprintf(pathInput, sizeof pathInput, "%s/%c", DEFAULT_PLAYERS,
		    letter1);
		snprintf(command, sizeof command, "ls -1 %s", pathInput);

		if ((fpPlayerList = popen(command, "r")) == NULL)
			continue;

		while (1) {
			fgets(e.name, MAX_LOGIN_NAME, fpPlayerList);

			if (feof(fpPlayerList))
				break;

			len = strlen(e.name);
			e.name[len - 1] = '\0';

			if (e.name[0] != letter1) {
				printf("File %c/%s: wrong directory.\n",
				    letter1, e.name);
			} else {
				snprintf(pathInput, sizeof pathInput,
				    "%s/%c/%s", DEFAULT_PLAYERS, letter1,
				    e.name);

				if (GetPlayerInfo(pathInput, &e)) {
					if ((list[n] = malloc(sizeof(ENTRY))) ==
					    NULL)
						err(1, "%s: malloc", __func__);

					memcpy(list[n], &e, sizeof(ENTRY));

					if (++n == listsize) {
						listsize += 100;
						list = realloc(list, listsize *
						    sizeof(ENTRY *));
						if (list == NULL) {
							err(1, "%s: realloc",
							    __func__);
						}
					}
				}
			}
		} /* while (1) */

		pclose(fpPlayerList);
	} /* for */

	return n;
}

static int
SetComputers(int n)
{
	FILE	*fpComp;
	char	 comp[30];
	char	 line[100];
	int	 i = 0;

	if (snprintf(line, sizeof line, "sort -f %s", COMPUTER_FILE) >=
	    sizeof line) {
		warnx("%s: snprintf truncated", __func__);
		return 0;
	} else if ((fpComp = popen(line, "r")) == NULL)
		return 0;

	while (i < n) {
		fgets(comp, 29, fpComp);

		if (feof(fpComp))
			break;

		comp[strlen(comp) - 1] = '\0';

		while (i < n && strcasecmp(list[i]->name, comp) < 0)
			i++;
		if (i < n && strcasecmp(list[i]->name, comp) == 0)
			list[i++]->computer = 1;
	}

	pclose(fpComp);
	return 1;
}

static int
sortfunc(const void *i, const void *j)
{
	int n =	(*(ENTRY **)j)->r[rtype].rating -
		(*(ENTRY **)i)->r[rtype].rating;

	return n ? n : strcasecmp((*(ENTRY **)i)->name, (*(ENTRY **)j)->name);
}

static void
makerank(void)
{
	FILE	*fp;
	char	 fName[200];
	int	 sortnum, sortmesize, i, n;

	printf("Loading players\n");
	n = LoadEntries();
	printf("Found %d players.\n", n);
	printf("Setting computers.\n");
	SetComputers(n);

	for (rtype = 0; rtype < 4; rtype++) {
		sortnum		= 0;
		sortmesize	= 100;
		sortme		= malloc(sizeof(ENTRY *) * sortmesize);

		if (sortme == NULL)
			err(1, "%s: malloc", __func__);

		for (i = 0; i < n; i++) {
			if (list[i]->r[rtype].rating) {
				sortme[sortnum++] = list[i];

				if (sortnum == sortmesize) {
					sortmesize += 100;
					sortme = realloc(sortme, sortmesize *
					    sizeof(ENTRY *));
					if (sortme == NULL)
						err(1, "%s: realloc", __func__);
				}
			}
		}

		printf("Sorting %d %s.\n", sortnum, rnames[rtype]);
		qsort(sortme, sortnum, sizeof(ENTRY *), sortfunc);

		printf("Saving to file.\n");
		snprintf(fName, sizeof fName, "%s/rank.%s", DEFAULT_STATS,
		    rnames[rtype]);

		if ((fp = fopen(fName, "w")) == NULL)
			err(1, "%s: fopen", __func__);

		for (i = 0; i < sortnum; i++) {
			fprintf(fp, "%s %d %d %d\n",
			    sortme[i]->name,
			    sortme[i]->r[rtype].rating,
			    sortme[i]->r[rtype].num,
			    sortme[i]->computer);
		}
		fclose(fp);
		free(sortme);
	}
}

int
main(int argc, char **argv)
{
	if (argc > 1) {
		fprintf(stderr, "usage: %s.\n", argv[0]);
		return EXIT_FAILURE;
	}

	makerank();
	return EXIT_SUCCESS;
}
